### PR TITLE
Fix/mathematical calculator

### PR DESCRIPTION
**Pull request by sjd9021**
> [!IMPORTANT]
> Replaces `eval` with a safe AST-based evaluation in `Calculator` to prevent arbitrary code execution and improve error handling.
> 
>   - **Security**:
>     - Replaces `eval` with AST-based evaluation in `Calculator.execute()` to prevent arbitrary code execution.
>   - **Functionality**:
>     - Introduces `_safe_eval()` in `calculator.py` to evaluate AST nodes, supporting `Add`, `Sub`, `Mult`, `Div`, `Pow`, `USub`, and `UAdd` operations.
>     - Handles `Constant`, `BinOp`, and `UnaryOp` node types in `_safe_eval()`.
>   - **Error Handling**:
>     - Returns error messages in `CalculatorResponse` if evaluation fails.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ComposioHQ%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for 7cd67b3d700b4433526e60ffa552aa6771824ac6. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/composio/tools/local/mathematical/actions/calculator.py [hunk 1]
```diff
@@ -19,12 +21,54 @@ class CalculatorResponse(BaseModel):
...
+        ast.Add: operator.add,
+        ast.Sub: operator.sub,
+        ast.Mult: operator.mul,
+        ast.Div: operator.truediv,
+        ast.Pow: operator.pow,
+        ast.USub: operator.neg,
+        ast.UAdd: operator.pos,
+    }
+
     def execute(self, request: CalculatorRequest, metadata: Dict) -> CalculatorResponse:
```
Comment (shreysingla11) [inline python/composio/tools/local/mathematical/actions/calculator.py:40]: Great job on addressing the security vulnerability by replacing `eval()` with a safer `ast`-based implementation. This significantly reduces the risk of code injection attacks. The new code structure using `ast` is more robust and easier to maintain.

File: python/composio/tools/local/mathematical/actions/calculator.py [hunk 2]
```diff
@@ -19,12 +21,54 @@ class CalculatorResponse(BaseModel):
...
-        return CalculatorResponse(
-            result=str(eval(request.operation))  # pylint: disable=eval-used
-        )
+        try:
+            # Parse the expression into an AST
+            node = ast.parse(request.operation, mode="eval").body
+            result = self._safe_eval(node)
+            return CalculatorResponse(result=str(result))
+        except Exception as e:
```
Comment (shreysingla11) [inline python/composio/tools/local/mathematical/actions/calculator.py:46]: Good addition of error handling. This will make debugging easier and provide more informative responses to users when invalid expressions are entered. Consider adding more specific error messages for different types of exceptions to further improve user experience.

File: python/composio/tools/local/mathematical/actions/calculator.py [hunk 3]
```diff
@@ -19,12 +21,54 @@ class CalculatorResponse(BaseModel):
 
 class Calculator(LocalAction[CalculatorRequest, CalculatorResponse]):
     """
-    Useful to perform any mathematical calculations, like sum, minus, multiplication, division, etc.
+    Performs mathematical calculations such as addition, subtraction, multiplication, and division.
```
Comment (shreysingla11) [inline python/composio/tools/local/mathematical/actions/calculator.py:24]: Consider expanding the class docstring to include information about supported operations and any limitations of the calculator. This would provide better documentation for users of the Calculator class.

File: python/composio/tools/local/mathematical/actions/calculator.py [hunk 4]
```diff
@@ -19,12 +21,54 @@ class CalculatorResponse(BaseModel):
...
+            # Parse the expression into an AST
+            node = ast.parse(request.operation, mode="eval").body
+            result = self._safe_eval(node)
+            return CalculatorResponse(result=str(result))
+        except Exception as e:
+            # Handle exceptions and return error messages
+            return CalculatorResponse(result=f"Error: {str(e)}")
+
+    def _safe_eval(self, node):
+        if isinstance(node, ast.Constant):  # <number>
```
Comment (shreysingla11) [inline python/composio/tools/local/mathematical/actions/calculator.py:51]: The `_safe_eval` method is handling multiple concerns. Consider refactoring it into smaller methods, each handling a specific type of AST node. This could improve readability and make the code easier to test. For example, you could have separate methods for evaluating constants, binary operations, and unary operations.

**Comment by shreysingla11** [comment]
Overall, this PR significantly improves the security and robustness of the Calculator class. The replacement of `eval()` with an `ast`-based implementation is a crucial security enhancement. The code structure is more maintainable, and the addition of error handling is beneficial.

Here's a summary of the improvements and suggestions:

1. Security: Great improvement by replacing `eval()` with `ast`.
2. Error Handling: Good addition, consider adding more specific error messages.
3. Code Structure: The new implementation is more robust, but consider further refactoring of the `_safe_eval` method.
4. Documentation: Expand the class docstring to include supported operations and limitations.
5. Testing: Ensure comprehensive unit tests are added to cover various scenarios and edge cases.

While there's room for some minor improvements in documentation and code organization, this PR represents a significant step forward in terms of security and functionality. Great work!
